docs/dev_plan.md

```diff
@@ -0,0 +1,339 @@
+expensive mathematical operations into an aggregate form, and structures
+them so that performance libraries or compilers can parallelize the
+computation. This pattern helps to encapsulate expensive operations and
+algorithms. Additionally, it follows the modular architecture design
+described above in that it supports swapping accelerators or parallelization
+methods../
+
+
+#### Data-oriented design
+
+OpenTurbine developers designing new algorithms and data structures
+should become familiar with the concepts of [data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design),
+particularly [structures of arrays vs. arrays of structures](https://stackoverflow.com/questions/17924705/structure-of-arrays-vs-array-of-structures).
+The key concept of this paradigm is to structure data so that it maps
+closely to the form it will be represented and used within the relevant
+processing units. For example, it is a common mistake in object-oriented
+design to represent data as arrays of structures (AoS),
+and this can lead to inefficiency in CPU cache
+usage since aggregate operations such as matrix math must load
+data from non-contiguous memory rather than perform the operations
+in a vector form. In a data-oriented design, data is more
+commonly represented as structures of arrays (SoA) where contiguous
+portions of memory are loaded into the cache and operated on
+either by the CPU or a hardware accelerator. The graphic below
+illustrates the difference between structures of arrays and arrays
+of structures.
+
+```{image} images/AoS_SoA.pdf
+:alt: aos_soa
+:width: 400px
+:align: center
+```
+
+The difference in usage for the above two patterns is illustrated via
+an example Python-based math operation below where a series of points
+are shifted in one direction by a constant amount.
+
+```python
+class Point:
+    x: int
+    y: int
+    z: int
+
+# Initialize test data
+all_points = [Point(random(i), random(i), random(i)) for i in range(100)]
+
+# Array of Structures
+AOS = [all_points[i].y for i in range(100)]
+
+# Structure of Arrays
+SOA = all_points.y + 10
+```
+
+While the AoS vs SoA pattern is a typical example of data-oriented vs object-oriented modeling,
+the concept extends to all areas of data modeling in software and is equally relevant to most
+programming languages.
+
+### Accessible software
+
+Access to the software is an important consideration for the
+longevity and relevance of OpenTurbine. In short, if the software is
+not accessible, it won't be used, extended, or maintained. All
+development efforts should always consider user and developer accessibility
+as a key driver. The distinction between "user" and "developer"
+is not always clear, but accessibility efforts should address concerns
+for both types of engagement with OpenTurbine.
+
+Documentation is the primary tool for addressing accessibility. As a
+guideline, the practice of documentation driven development should be
```

Comment:
Somewhere else in the document it said we would use test driven development. Documentation driven development is not the same and is vastly inferior.

Reply:
In this case, I intended that documentation driven development be adopted in addition to TDD rather than instead of it. I agree with your comment below that unit tests are the best way to document low-level code. For high-level code and public API's, though, I think we need to strike a balance of testing plus written documentation even at the risk of documentation becoming out of date. I see this as critical to ensuring that early design decisions and concepts are communicated to future developers so that they may understand when they're making fundamental changes. It also reduces the barrier to entry for future developers. Ultimately, the burden of ensuring that this documentation is accurate is on the development team who ideally treat the docs as a first-priority just like the code.

Reply:
Great points, thanks for the extra explanation. I totally agree that documentation is valuable, especially for high-level concepts and public APIs. It's definitely true that the developers need to prioritize keeping the docs up to date, because it doesn't happen automatically in the way that code and tests keep each other up to date.